src/dir.targets

```diff
@@ -178,6 +203,28 @@
     <Error Condition="'$(TestRunExitCode)' != '0'" Text="One or more tests failed while running tests from '$(MSBuildProjectName)' please check log for details!" />
   </Target>
 
+  <Target Name="CheckTestCategories"
+          AfterTargets="Test"
+          Condition="'$(TestDisabled)'!='True'">
+
+    <ItemGroup>
+      <!-- Intersection of TestCategoriesItems and RunTestsWithCategoriesItems -->
+      <RunTestsCategoriesFiltered Include="@(RunTestsWithCategoriesItems)" Condition="'@(RunTestsWithCategoriesItems)'=='@(TestCategoriesItems)' And '%(Identity)'!=''" />
+      <RunTestsCategoriesFiltered Remove="@(RunTestsWithoutCategoriesItems)" />
```

Comment:
We want an intersection here. If I remove it then I will get subtract(A, B) instead of intersect(A, B).

This condition does count(A)*count(B) operation and is being expanded for each combination and if element exists in both sets then is being added to the resulting set. This only works because of "'%(Identity)'!=''" which causes it to enumerate. Without that it would simply do:
'JOIN(A,";")'=='JOIN(B,";")' which would not work. In addition to that it removes empty elements which we want too.